File: exp07_rotation.cpp
// ===================================================================
// Experiment 07: 2D Rotation of a Line and Triangle
//
// THEORY:
// Rotation turns an object around a fixed point (called the pivot).
// By convention, positive angle = counterclockwise rotation.
//
// FORMULA (rotation about origin by angle θ):
//   x' = x·cos(θ) - y·sin(θ)
//   y' = x·sin(θ) + y·cos(θ)
//
// MATRIX FORM:
//   | x' |   | cosθ  -sinθ  0 |   | x |
//   | y' | = | sinθ   cosθ  0 | * | y |
//   | 1  |   |  0      0    1 |   | 1 |
//
// ROTATION ABOUT ARBITRARY POINT (cx, cy):
// We can't just rotate around any point with the simple formula.
// We need 3 steps:
//   1. TRANSLATE so the pivot goes to origin:  T(-cx, -cy)
//   2. ROTATE about origin:                    R(θ)
//   3. TRANSLATE back:                         T(cx, cy)
//
// Combined:  P' = T(cx, cy) · R(θ) · T(-cx, -cy) · P
//
// In code:  x_rel = x - cx,  y_rel = y - cy
//           x_new = cx + x_rel·cosθ - y_rel·sinθ
//           y_new = cy + x_rel·sinθ + y_rel·cosθ
// ===================================================================

#include <graphics.h>
#include <cmath>         // For cos(), sin()
#include <conio.h>
#include <cstdio>

#define PI 3.14159265    // π for converting degrees to radians

// Helper: Draw a triangle
void drawTriangle(int x1, int y1, int x2, int y2, int x3, int y3, int color) {
    setcolor(color);
    line(x1, y1, x2, y2);
    line(x2, y2, x3, y3);
    line(x3, y3, x1, y1);
}

// Helper: Draw a line
void drawLine(int x1, int y1, int x2, int y2, int color) {
    setcolor(color);
    line(x1, y1, x2, y2);
}

// Function: Rotate a point (x,y) about pivot (cx,cy) by angle degrees
// Result is stored in (xOut, yOut)
void rotatePoint(int x, int y, int cx, int cy, float angle, int &xOut, int &yOut) {
    // Step 1: Convert angle from degrees to radians
    float rad = angle * PI / 180.0;
    float cosA = cos(rad);
    float sinA = sin(rad);

    // Step 2: Translate point relative to pivot
    // (xRel, yRel) = position relative to rotation center
    int xRel = x - cx;
    int yRel = y - cy;

    // Step 3: Apply rotation formula about origin
    // x_new = x·cosθ - y·sinθ
    // y_new = x·sinθ + y·cosθ
    int xRot = (int)(xRel * cosA - yRel * sinA + 0.5f);
    int yRot = (int)(xRel * sinA + yRel * cosA + 0.5f);

    // Step 4: Translate back to original position
    xOut = cx + xRot;
    yOut = cy + yRot;
}

// ===================================================================
// MAIN FUNCTION
// ===================================================================
int main() {
    int gd = DETECT, gm;
    initgraph(&gd, &gm, "");

    // ================================================================
    // PART 1: Rotate a LINE about its start point
    // ================================================================
    int lx1 = 100, ly1 = 300;   // Pivot point (start of line)
    int lx2 = 250, ly2 = 300;   // End of line

    // Original line (horizontal, to the right)
    drawLine(lx1, ly1, lx2, ly2, WHITE);
    outtextxy(lx1, ly1 + 10, "Original");

    // Rotate by 45 degrees
    int rx, ry;
    rotatePoint(lx2, ly2, lx1, ly1, 45, rx, ry);
    drawLine(lx1, ly1, rx, ry, YELLOW);
    outtextxy(rx + 5, ry, "Rotated 45 deg");

    // Rotate by 90 degrees
    rotatePoint(lx2, ly2, lx1, ly1, 90, rx, ry);
    drawLine(lx1, ly1, rx, ry, GREEN);
    outtextxy(rx + 5, ry, "Rotated 90 deg");

    // ================================================================
    // PART 2: Rotate a TRIANGLE about its centroid
    // ================================================================
    // Triangle vertices
    int x1 = 400, y1 = 80;      // Top
    int x2 = 500, y2 = 200;     // Right
    int x3 = 300, y3 = 200;     // Left

    // Centroid = average of all vertices
    // This is the "center of mass" of the triangle
    int cx = (x1 + x2 + x3) / 3;
    int cy = (y1 + y2 + y3) / 3;

    // Original triangle
    drawTriangle(x1, y1, x2, y2, x3, y3, WHITE);
    outtextxy(cx - 30, cy, "Original");

    // Rotated by 60 degrees
    int rx1, ry1, rx2, ry2, rx3, ry3;
    rotatePoint(x1, y1, cx, cy, 60, rx1, ry1);
    rotatePoint(x2, y2, cx, cy, 60, rx2, ry2);
    rotatePoint(x3, y3, cx, cy, 60, rx3, ry3);
    drawTriangle(rx1, ry1, rx2, ry2, rx3, ry3, MAGENTA);
    outtextxy(cx - 30, cy + 80, "Rotated 60 deg");

    // Rotated by 120 degrees
    rotatePoint(x1, y1, cx, cy, 120, rx1, ry1);
    rotatePoint(x2, y2, cx, cy, 120, rx2, ry2);
    rotatePoint(x3, y3, cx, cy, 120, rx3, ry3);
    drawTriangle(rx1, ry1, rx2, ry2, rx3, ry3, CYAN);
    outtextxy(cx - 30, cy + 160, "Rotated 120 deg");

    outtextxy(10, 10, "Experiment 07: 2D Rotation");
    outtextxy(10, 30, "Rotation about line-start / triangle-centroid");

    getch();
    closegraph();
    return 0;
}

File: exp07_rotation.py
import math
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches


def rotate_point(x, y, cx, cy, angle):
    rad = math.radians(angle)
    cos_a, sin_a = math.cos(rad), math.sin(rad)
    x_rel, y_rel = x - cx, y - cy
    x_rot = x_rel * cos_a - y_rel * sin_a
    y_rot = x_rel * sin_a + y_rel * cos_a
    return round(cx + x_rot), round(cy + y_rot)


def draw_triangle(ax, x1, y1, x2, y2, x3, y3, color, label=None):
    tri = mpatches.Polygon(
        [(x1, y1), (x2, y2), (x3, y3)],
        fill=False, edgecolor=color, linewidth=2, label=label
    )
    ax.add_patch(tri)


def run():
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5.5))
    fig.suptitle("2D Rotation Transformation", fontsize=13)

    # ---- Part 1: Rotate a LINE ----
    ax1.set_title("Rotation: Line")
    ax1.set_xlim(50, 350)
    ax1.set_ylim(200, 380)
    ax1.set_aspect("equal")
    ax1.grid(True, alpha=0.3)

    lx1, ly1 = 100, 300
    lx2, ly2 = 250, 300

    ax1.plot([lx1, lx2], [ly1, ly2], "w-", linewidth=2, label="Original")
    ax1.scatter([lx1], [ly1], c="red", s=50, zorder=4)

    for deg, color, label in [(45, "yellow", "Rotated 45"), (90, "lime", "Rotated 90")]:
        rx, ry = rotate_point(lx2, ly2, lx1, ly1, deg)
        ax1.plot([lx1, rx], [ly1, ry], color=color, linewidth=2, label=label)
        ax1.scatter([rx], [ry], c=color, s=40, zorder=4)
        ax1.annotate(label, (rx, ry), xytext=(5, 5),
                     textcoords="offset points", color=color, fontsize=8)

    ax1.annotate("Pivot", (lx1, ly1), xytext=(5, 8),
                 textcoords="offset points", color="red", fontsize=8)

    # ---- Part 2: Rotate a TRIANGLE ----
    ax2.set_title("Rotation: Triangle")
    ax2.set_xlim(200, 600)
    ax2.set_ylim(0, 350)
    ax2.set_aspect("equal")
    ax2.grid(True, alpha=0.3)

    x1, y1 = 400, 80
    x2, y2 = 500, 200
    x3, y3 = 300, 200
    cx = (x1 + x2 + x3) // 3
    cy = (y1 + y2 + y3) // 3

    draw_triangle(ax2, x1, y1, x2, y2, x3, y3, "white", "Original")
    ax2.scatter([cx], [cy], c="red", s=50, zorder=4)
    ax2.annotate("Centroid", (cx, cy), xytext=(5, 8),
                 textcoords="offset points", color="red", fontsize=8)

    for deg, color, label in [(60, "magenta", "Rotated 60"), (120, "cyan", "Rotated 120")]:
        r = [rotate_point(x1, y1, cx, cy, deg),
             rotate_point(x2, y2, cx, cy, deg),
             rotate_point(x3, y3, cx, cy, deg)]
        draw_triangle(ax2, r[0][0], r[0][1], r[1][0], r[1][1],
                      r[2][0], r[2][1], color, label)
        ax2.scatter([r[0][0]], [r[0][1]], c=color, s=30, zorder=4)

    for ax in [ax1, ax2]:
        ax.set_xlabel("X")
        ax.set_ylabel("Y")
        ax.legend(fontsize=7, loc="lower right")
        ax.set_facecolor("#2b2b2b")
        ax.tick_params(colors="white")
        ax.xaxis.label.set_color("white")
        ax.yaxis.label.set_color("white")
        ax.title.set_color("white")

    fig.patch.set_facecolor("#2b2b2b")
    plt.tight_layout()
    plt.show()


if __name__ == "__main__":
    run()
